templates/consul.sysv.erb

```diff
@@ -61,18 +61,42 @@ start() {
 }
 
 stop() {
+        DELAY=5 # seconds maximum to wait for a leave
+
         echo -n "Shutting down consul: "
+        mkpidfile
+
         # If consul is not acting as a server, exit gracefully
+        # Use SIGINT to create a "leave" event, unless the user has explicitly
+        # changed that behavior in the Consul config.
         if ("${CONSUL}" info ${RPC_ADDR} 2>/dev/null | grep -q 'server = false' 2>/dev/null) ; then
-            "$CONSUL" leave ${RPC_ADDR}
-        fi
+            consul_pid=$(cat $PID_FILE)
+            killproc $KILLPROC_OPT $CONSUL -INT
+            retcode=$?
 
-        # If acting as a server, or if leave failed, kill it.
-        mkpidfile
-        killproc $KILLPROC_OPT $CONSUL -9
+            # We'll wait if necessary to make sure the leave works, and return
+            # early if we can.  If not, escalate to harsher signals.
+            try=0
+            while [ $try -lt $DELAY ]; do
```

Comment:
One thing we need to consider is what would happen if the PID never dies and `$try` is greater than or equal to `$DELAY`. At that point it will break out of the loop and execute `-TERM`. I don't think thats desired?